draft-irtf-cfrg-vdaf.md

```diff
@@ -603,83 +795,32 @@ privacy).
 
 ## Aggregation {#sec-vdaf-aggregate}
 
-Once an Aggregator holds output shares for a batch of measurements (where
-batches are defined by the application), it combines them into a share of the
-desired aggregate result.  This algorithm is performed locally at each
-Aggregator, without communication with the other Aggregators.
+VDAF Aggregation is identical to DAF Aggregation (cf. {{sec-daf-aggregate}}):
 
-* `Vdaf.out_shares_to_agg_share(agg_param: AggParam, output_shares:
-  Vec[OutShare]) -> agg_share: AggShare` is the deterministic aggregation
-  algorithm. It is run by each Aggregator over the output shares it has computed
-  over a batch of measurement inputs.
+* `Vdaf.out_shares_to_agg_share(agg_param: AggParam, out_shares: Vec[OutShare])
+  -> agg_share: AggShare` is the deterministic aggregation algorithm. It is run
+  by each Aggregator over the output shares it has computed over a batch of
+  measurement inputs.
 
-~~~~
-    Aggregator 0    Aggregator 1        Aggregator SHARES-1
-    ============    ============        ===================
-
-    out_share_0_0   out_share_1_0       out_share_[SHARES-1]_0
-    out_share_0_1   out_share_1_1       out_share_[SHARES-1]_1
-    out_share_0_2   out_share_1_2       out_share_[SHARES-1]_2
-         ...             ...                     ...
-    out_share_0_B   out_share_1_B       out_share_[SHARES-1]_B
-      |               |                   |
-      V               V                   V
-    +-----------+   +-----------+       +-----------+
-    | out2agg   |   | out2agg   |   ... | out2agg   |
-    +-----------+   +-----------+       +-----------+
-      |               |                   |
-      V               V                   V
-    agg_share_0     agg_share_1         agg_share_[SHARES-1]
-~~~~
-{: #agg-flow title="Aggregation of output shares. `B` indicates the number of
-measurements in the batch."}
-
-For simplicity, we have written this algorithm and the unsharding algorithm
-below in "one-shot" form, where all shares for a batch are provided at the same
-time.  Some VDAFs may also support a "streaming" form, where shares are
-processed one at a time.
+The data flow for this stage is illustrated in {{aggregate-flow}}. Here again,
+we have the aggregation algorithm in a "one-shot" form, where all shares for a
+batch are provided at the same time. VDAFs typically also support a "streaming"
+form, where shares are processed one at a time.
```

Comment:
I'm a bit confused about "one-shot" vs. "streaming". The distinction that I'm aware of between VDAFs is whether the aggregation parameter exists or not. If it does, then aggregation can't start until the collector proves the aggregation parameter. If it does not, then aggregators can aggregate shares immediately upon upload.

But I don't think that has anything to do with whether aggregation can be "streamed". Let's say we're running `poplar1`, and my aggregator receives 50 shares, and then the collector sends me `agg_param`. Now I can aggregate together those 50 shares, but I'm likely to do so by preparing them individually, independently from each other, summing each output share into an aggregate, which sounds like "streaming" to me. Further, my aggregator may continue to receive shares between when it receives `agg_param` (in a DAP `CollectReq`) and when it delivers the aggregate share (which will happen when the collector hits the DAP collect job URI), and I don't think anything stops the aggregator from continuing to incorporate shares into the aggregate -- or _stream_ them.

I think I know what you mean by "one-shot" vs. "streaming", but I don't think these are the right words to express it, since the issue here is when the aggregator can expect to have the aggregation parameter and not whether input shares can be aggregated incrementally.

Reply:
Both `prio3` and `poplar1` support "streaming" aggregation in the sense used here, as both there aggregation operations are effectively vector addition in a vector space over a field. The distinction is that `out_shares_to_agg_share` as given here takes in all output shares at once, and returns an aggregate share, rather than processing output shares only one at a time, updating an accumulator, and returning an aggregate share at the end. For a strawman counterexample, consider a VDAF where `out_shares_to_agg_share` takes the median of numbers from each output share. This would be a non-streaming VDAF, as there's no efficient way to incrementally compute an aggregate share. Instead, the only way to compute the exact median is to keep the multiset of all output shares around until all output shares are received.

Reply:
Maybe this is too big of a nomenclature issue to tackle in this particular PR

Reply:
It's arguably orthogonal given that "one-shot" is the current terminology (i.e., not introduced by this PR). How about adding a comment to #47 about fixing this?